rfcs/0144-versioned-flake-references.md

```diff
@@ -0,0 +1,120 @@
+---
+feature: versioned-flake-references
+start-date: 2023-03-19
+author: figsoda
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Flake references can have a special placeholder that will specify the version
+requirement of the flake, which `nix flake update` can use to update the pin to
+the latest version compatible with the version specification. A new Nix
+command, `nix flake upgrade` will upgrade the version requirement in
+`flake.nix` to the latest possible version without regard to compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+This will allow Nix libraries to be versioned without requiring their users
+to manually update them. Some package managers (e.g. cargo) for more
```

Comment:
Here you cite cargo, but the specification they employ to do those machine-driven updates is very dumb - by design. Cargo can update patch version, but minor version has restrictions and major is almost impossible.

It looks so full of restrictions that it is easier to edit by hand.

Reply:
Could you elaborate on what "full of restrictions" means? I don't think attack my references is a good explanation.

Reply:
The example tool you gave, Cargo, is purposefully very limited. This design decision conveys the idea that major and minor versions should not be upgraded by a machine, but by a human being.

But your specification is too complex (maybe a consequence of Nix flakes being too complex), and it feels problematic to me.

Reply:
> This design decision conveys the idea that major and minor versions should not be upgraded by a machine, but by a human being.

The whole point of defining a set of rules for the version specification is to make them upgradable by a machine, hence the strict rules on what version bumps can introduce breaking changes. This is so libraries can follow a predictable pattern, instead of their own preference, so updating can be automated.

Your response still doesn't explain what is limited, and I don't see why Cargo being supposedly limited has anything to do with my proposal being problematic. The purpose of this proposal is to give library authors the freedom to version their library, instead of the current situation of users following the main branch and library authors not having a good communicable way to introduce breaking changes.

Reply:
> Your response still doesn't explain what is limited

I said, and I will say again: Cargo (or more precisely `cargo.toml`) is limited. Being limited, it is easy to understand and implement.

Your proposal is too wide. It introduces more syntatical and semantical rules - implying more burden for the compiler writer (of a language with no formal specification yet, by the way).

Further,

> give library authors the *freedom* to version their library

Goes against

> *strict rules* on what version bumps can introduce breaking changes.

The design decision should find the equilibrium between these two points.

Reply:
> Cargo (or more precisely cargo.toml) is limited.

If you are talking about the expressiveness of `Cargo.toml` vs `flake.nix`, I agree. However, I don't think the difference applies to this specific feature.

> It introduces more syntatical and semantical rules.

Cargo defines [much](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html) [more](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility) [rules](https://doc.rust-lang.org/cargo/reference/semver.html) in this aspect and also more strictly enforces it, as it will fail if the version is absent or not a valid semver.

> Implying more burden for the compiler writer

I agree.

> **freedom** goes against **strict rules**

It doesn't. They are talking about completely different things. This feature will always be opt-in. Library authors can still not version their libraries, or define their own versioning scheme - they just wouldn't be benefited by the change. The only way tooling like this can happen is if people following a predictable set of rules when versioning their libraries, which this proposal will allow by defining strict rules.